### ISSUE TITLE
Yankee Trader "freezes" when starting up

### ISSUE THREAD

**Issue by nelgin** [dosemu2/dosemu2#1706]
**Describe the bug**
Yankee Trader is a DOS door for BBS systems. After configuring my environment and testing it with another door to confirm it works, I ran Yankee Trader and it just hangs. Debugging shows something happening but I cannot tell what.

**To Reproduce**
Install VirtualBox on Windows 10, Install Ubuntu 20.04 in VB. Download and configure Synchronet BBS within Ubuntu. Install DOSEMU2 from ppa. Download and install Yankee Trader and configure it then run it via the BBS door menu.

**Attach program/game binaries or provide an URL**
https://yt.starflt.com/yt/Install/YT36d.ZIP

**Attach the log**
It is located in ~/.dosemu/boot.log
[dosemu_boot.log](https://github.com/dosemu2/dosemu2/files/8998473/dosemu_boot.log)

**Additional info**
The DOSEMU command run by the BBS is:
/usr/bin/env  /usr/bin/dosemu.bin -q -td -I"video { none }" -I"serial { virtual com 1 }" -f/sbbs/ctrl/dosemu.conf -d /sbbs/node1/ -d /sbbs/xtrn -d /sbbs/ctrl/ -d /sbbs/data/ -d /sbbs/exec/ -E D:external.bat -o/sbbs/node1/dosemu_boot.log -D+a

Also attached is external.bat which the dosemu command runs. It basically changes to the directory where the door game is, loads in a fossil driver if it finds one (x00.exe in my case) and then run the game.

[external.bat.txt](https://github.com/dosemu2/dosemu2/files/8998506/external.bat.txt)

If I run other programs such as YTCONFIG then it also hangs unless I remove the -td - I tried this via the BBS but it hangs too.

I'm happy to provide a screenshare or capture a video which explains what's going on, since this isn't your normal "run from the command line dosemu" situation.

**Comment by andrewbird** [comment]
Do you have your /sbbs/ctrl/dosemu.conf please?

**Comment by stsp** [comment]
> If I run other programs such as YTCONFIG then it also hangs unless I remove the -td - I tried this via the BBS but it hangs too.

Of course it won't work in -td
mode. Use -t or gfx backends.

**Comment by stsp** [comment]
In fact, using `-I"serial { virtual com 1 }`
with -t/-td is a very bad idea because
you don't see a boot process.
So at least for debugging you can try
the gfx back-end.

**Comment by stsp** [comment]
I added the change that allows
to see boot process if you use
virtual com with `-tde`.

**Comment by andrewbird** [comment]
I noticed this `cannot be requested` in the log.
~~~
SER0: INT14 0x0: Initialize port 0, AL=0xa3                                     
SER0: Running ser_open, /dev/pts/3, fd=-1                                       
Opening Virtual Port                                                            
SER0: do_ser_init: running ser_termios                                          
SER0: LCR = 0x3, 8N1, bps = 2400, divisor 0x30 -> 0x30                          
GEN: fd=39 gets SIGIO for async_serial_run                                      
SER0: Modem Status Change: MSR -> 0xb0                                          
SER0: Func modstat_engine requesting MS_INTR                                    
SER0: Interrupt 4 (8) cannot be requested: enable=0 IER=0x0                     
SER0: Read LCR = 0x3                                                            
SER0: LCR = 0x83, DLAB high.                                                    
SER0: Interrupt 4 (8) cannot be requested: enable=0 IER=0x0                     
SER0: Divisor LSB = 0x30                                                        
SER0: Interrupt 4 (8) cannot be requested: enable=0 IER=0x0                     
SER0: Divisor MSB = 0x0                                                         
SER0: Interrupt 4 (8) cannot be requested: enable=0 IER=0x0                     
SER0: LCR = 0x3, DLAB low.                                                      
SER0: Clear FIFO.                                                               
SER0: LCR = 0x3, 8N1, bps = 2400, divisor 0x30 -> 0x30                          
SER0: Interrupt 4 (8) cannot be requested: enable=0 IER=0x0                     
SER0: Read MCR = 0x0                                                            
SER0: MCR = 0x0 -> 0x3                                                          
SER0: MCR: DTR -> 1                                                             
SER0: MCR: RTS -> 2                                                             
SER0: Interrupt 4 (8) cannot be requested: enable=0 IER=0x0                     
SER0: Read LSR = 0x60                                                           
SER0: Read MSR = 0xbb                                                           
SER0: INT14 0x0: Return with AL=0xbb AH=0x60                                    
~~~

Does that look okay?

**Comment by stsp** [comment]
This just means that the interrupts
were not enabled in the appropriate
registers, so they won't trigger now.
I don't think it makes any sense to
look into a log just yet. What you see
is perhaps the external fossil
initialization. We don't know if the
main prog was started at all because
you can't see the boot process w/o
the aforementioned patch.
The guy is trying to blindly convert
from dosemu1's config. This is a no
go. He needs to see what exactly
is going on, and then he will quickly
find the problem w/o any logs.

**Comment by nelgin** [comment]
Sorry, forgot to include the dosemu.conf 

$_cpu = "80486"
$_cpu_emu = "fullsim"
$_floppy_a = ""
$_cdrom = ""
$_xms = (1024)
$_ems = (1024)
$_ems_frame = (0xe000)
$_external_char_set = "cp437"
$_internal_char_set = "cp437"
$_term_updfreq = (8)
$_layout = "us"
$_rawkeyboard = (auto)
$_mouse_internal = (on)
$_mouse_dev = ""
$_joy_device = ""
$_lpt1 = ""
$_lp2 = ""
$_speaker = ""
$_sound = (off)
$_sb_dsp = ""
$_hdimage = "+0 -5 +1"

**Comment by nelgin** [comment]
I just updated and I see the following in the log. This also displays within the BBS (with the exception of a couple of lines) and hangs at the "Build 2.0pre9". There's nothing further in the log and I had to kill the process.

FDPP kernel 1.6 [GIT: ] (compiled Jun 27 2022)^M
Kernel compatibility 7.10 - clang - FAT32 support^M
^M
Written by Stas Sergeev, FDPP project.^M
Based on FreeDOS sources (C) Pasquale J. Villani and The FreeDOS Project.^M
^M
This program is free software: you can redistribute it and/or modify^M
it under the terms of the GNU General Public License as published by^M
the Free Software Foundation, either version 3 of the License, or^M
(at your option) any later version.^M
^M
 - InitDisk: ^MC: HD1, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB^M
^MI: HD2, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB^M
^MJ: HD3, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB^M
^MK: HD4, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB^M
CONF: reserving 64Kb at 0xA0000 for 'U' (Upper Memory Block (UMB, XMS 3.0))
CONF: reserving 124Kb at 0xC1000 for 'U' (Upper Memory Block (UMB, XMS 3.0))
CONF: reserving 40Kb at 0xF0000 for 'U' (Upper Memory Block (UMB, XMS 3.0))
^M                                                                               ^Mdosemu XMS 3.0 & UMB support enabled^M
CONF: reserving 16Kb at 0xE0000 for 'E' (EMS page frame)
CONF: reserving 16Kb at 0xE4000 for 'E' (EMS page frame)
CONF: reserving 16Kb at 0xE8000 for 'E' (EMS page frame)
CONF: reserving 16Kb at 0xEC000 for 'E' (EMS page frame)
dosemu EMS driver rev 0.9 installed.^M
EMUFS host file and print access available^M
dosemu CDROM driver installed (V0.2)^M
Kernel: allocated 115 Diskbuffers = 62560 Bytes in HMA^M
Process 0 starting: J:\command.com /e:384 /p /k %FDPP_AUTOEXEC%^M^M
^M
FreeDOS userspace tools can be installed with: insfdusr^M
Sound not enabled in config!^M
Welcome to dosemu2!^M
    Build 2.0pre9^M

**Comment by stsp** [comment]
Yes, that's "correct", -tde only works
up to the -E program... but I guess
this is not what you want.
So I added another tweak now.

**Comment by stsp** [comment]
You can now write `-tde2` to get
the full output.

**Comment by nelgin** [comment]
Ok. I'll give it a try once ppa is updated. Thanks.

**Comment by nelgin** [comment]
@andrewbird How often is ppa updated? Would it be better for me to "roll my own" from git while testing or should I keep waiting for ppa to catch up with the changes?

**Comment by andrewbird** [comment]
Difficult to say exactly, launchpad has two independent processes 'import' and 'build': last import was 1 hour ago, last build was 6 hours ago. Anyway I just manually started a build on there, so try apt update / upgrade in about 15 mins.

**Comment by nelgin** [comment]
Will do, cheers.

**Comment by nelgin** [comment]
@stsp I have just updated and the -t option is now -tde2

The result is the same. It gets no further than displaying "Build 2.0pre9" both in the log and within the bbs.

I'll attach the log for reference.
[dosemu_boot.log](https://github.com/dosemu2/dosemu2/files/9005726/dosemu_boot.log)

**Comment by stsp** [comment]
In this case its possible that
your prog doesn't produce any
output to screen.
Unfortunately we didn't log
the prog started with -E so I
added that now.
It will be possible to see in the
log what was started by -E.

**Comment by nelgin** [comment]
If I use this command from the linux command line

/usr/bin/env  /usr/bin/dosemu.bin -q  -f/sbbs/ctrl/dosemu.conf -d /sbbs/node1/ -d /sbbs/xtrn -d /sbbs/ctrl/ -d /sbbs/data/ -d /sbbs/exec/  -E D:external.bat

and a slightly modified external.bat to run LOCAL rather than YT.EXE then it opens a new dosemu window and runs fine. until I gets this error: dosemu.bin: /build/dosemu2-HaDZ0T/dosemu2-2.0~pre9-7504/build/../src/plugin/sdl/sdl.c:1645: void SDL_draw_text_cursor(void *, int, int, Bit8u, int, int, Boolean): Assertion `d.tex' failed.
Which is a different matter because I won't want to run it in an SDL window anyway.

So when I add the -tde or -tde2

It gets to run LOCAL and just hangs again.

I removed the @echo off, removed the >nul redirection from the x00.exe loader, and changed YT to LOCAL near the bottom in the external.bat

So yes, it definitely produces output.
![image](https://user-images.githubusercontent.com/6232589/176339640-72c33cab-cd83-485f-8fb5-4645a5510522.png)

**Comment by stsp** [comment]
So what is needed to reproduce the
problem? Your external.bat refers to
some EMUSETUP.BAT which you
didn't provide.
Also of course we don't have that
/sbbs stuff.
But I can just run LOCAL.EXE and
get the same screen you show above.

**Comment by stsp** [comment]
It would be good if you provide
the minimal, simple to setup example.
Dozens of bat files, custom configs
etc are not helping.

**Comment by stsp** [comment]
Especially the SDL crash would be
very interesting to see, even if it is
not needed for your setup. :)

**Comment by nelgin** [comment]
I just did an update, looks like only fdpp was updated, so if you made a change to dosemu2 itself then I'll have to wait for that to build, but I did retest and same result. Nothing on the screen after Build 2.0pre9.

There is no emusetup.bat in this config. That's a hold over from the old dosemu1 setup. The script skips to NOEMU: which is where the FOSSIL driver is loaded.

From a terminal run this:
/usr/bin/env  /usr/bin/dosemu.bin -q -tde2 -f/sbbs/ctrl/dosemu.conf -d /sbbs/node1/ -d /sbbs/xtrn -d /sbbs/ctrl/ -d /sbbs/data/ -d /sbbs/exec/ -E D:external.bat

With the attached external.bat which will run the local version and removes the redirections to nul etc.

[external.bat.txt](https://github.com/dosemu2/dosemu2/files/9012264/external.bat.txt)
[output.txt](https://github.com/dosemu2/dosemu2/files/9012277/output.txt)

and it hangs.

If I remove the -tde2 then try again, it runs and asks for my name. I enter something, it displays a bunch of ansi and then dies with:

dosemu.bin: /build/dosemu2-HaDZ0T/dosemu2-2.0~pre9-7504/build/../src/plugin/sdl/sdl.c:1645: void SDL_draw_text_cursor(void *, int, int, Bit8u, int, int, Boolean): Assertion `d.tex' failed.

I'm attaching the debug output with -D+a for your reference but, of course, this is a secondary issue to not having anything display at all.


[debug.log.gz](https://github.com/dosemu2/dosemu2/files/9012305/debug.log.gz)

**Comment by nelgin** [comment]
> It would be good if you provide
> the minimal, simple to setup example.
> Dozens of bat files, custom configs
> etc are not helping.

mkdir ~/.dosemu/drive_c/yt
copy the game files into ~/.dosemu/drive_c/yt
run
dosemu.bin -tde2 -f dosemu.conf (using the previous dosemu.conf I gave you)
run the LOCAL command
dosemu hangs.

I think that's about as simple as it gets. No drivers, no bat files, just you, dosemu and YT.

You can also try ytconfig which just runs the config utility without any part of the game needing to be configured. That just completely hangs and I cannot CTRL-C out of it.

**Comment by andrewbird** [comment]
> I just did an update, looks like only fdpp was updated, so if you made a change to dosemu2 itself then I'll have to wait for that to build, but I did retest and same result. 

I just manually started a build on dosemu2, so try apt update / upgrade in about 15 mins.

**Comment by nelgin** [comment]
Thank you Andrew. I just updated and, as expected, same result.

**Comment by stsp** [comment]
> Thank you Andrew. I just updated and, as expected, same result.

Nothing was suppose to change,
I only did a small logging tweak.

> dosemu.bin -tde2 -f dosemu.conf (using the previous dosemu.conf I gave you)
> run the LOCAL command

Sure because -td/tde/tde2 are
not for local mode. Use -t for
local mode, and no "hang" then.

**Comment by stsp** [comment]
That is to say I was planning to
rewrite virtual comport code from
zero, so that you could use -t
for remote mode and overall
have fewer caveats.
But for now use -t for local mode
and -tde2 for remote.

**Comment by stsp** [comment]
Can I reproduce SDL crash with
just local.exe?

**Comment by nelgin** [comment]
> Can I reproduce SDL crash with
> just local.exe?

Yes, you can - but like I said, that isn't my main concert, unless that issue is stopping it running.

I did see something else flash on the screen just before it crashed.  Just another info screen. This also appears briefly with YT via SDL, but blink and you miss it.


![image](https://user-images.githubusercontent.com/6232589/176573410-1dd89620-14d1-4faa-b73d-d28a4f0eba6c.png)

OK well, I don't really care about local mode so much, I have it configured. I still can't run it.

As of now, the game, when configured with -tde2, does NOT run via the BBS.

This is the command it's using.
6/30 01:38:59 term Node 1 <Nelgin> running external program: Yankee Trader
6/30 01:38:59 term Node 1 <Nelgin> Executing external: YT D:\DORINFO1.DEF
6/30 01:38:59 term Node 1 <Nelgin> Executing cmd-line: /usr/bin/env  /usr/bin/dosemu.bin -q -tde2 -I"video { none }" -I"serial { virtual com 1 }" -f/sbbs/ctrl/dosemu.conf -d /sbbs/node1/ -d /sbbs/xtrn -d /sbbs/ctrl/ -d /sbbs/data/ -d /sbbs/exec/ -E D:external.bat -o/sbbs/node1/dosemu_boot.log


This is what is now displayed in my logfile and to the BBS user. After that point, the game does not run.

6/30 01:38:59 term Node 1 <Nelgin> YT: FDPP kernel 1.6 [GIT: ] (compiled Jun 29 2022)
6/30 01:38:59 term Node 1 <Nelgin> YT: Kernel compatibility 7.10 - clang - FAT32 support
6/30 01:38:59 term Node 1 <Nelgin> YT: Written by Stas Sergeev, FDPP project.
6/30 01:38:59 term Node 1 <Nelgin> YT: Based on FreeDOS sources (C) Pasquale J. Villani and The FreeDOS Project.
6/30 01:38:59 term Node 1 <Nelgin> YT: This program is free software: you can redistribute it and/or modify
6/30 01:38:59 term Node 1 <Nelgin> YT: it under the terms of the GNU General Public License as published by
6/30 01:38:59 term Node 1 <Nelgin> YT: the Free Software Foundation, either version 3 of the License, or
6/30 01:38:59 term Node 1 <Nelgin> YT: (at your option) any later version.
6/30 01:38:59 term Node 1 <Nelgin> YT: - InitDisk: C: HD1, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
6/30 01:38:59 term Node 1 <Nelgin> YT: I: HD2, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
6/30 01:38:59 term Node 1 <Nelgin> YT: J: HD3, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
6/30 01:38:59 term Node 1 <Nelgin> YT: K: HD4, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
6/30 01:38:59 term Node 1 <Nelgin> YT: dosemu XMS 3.0 & UMB support enabled
6/30 01:38:59 term Node 1 <Nelgin> YT: dosemu EMS driver rev 0.9 installed.
6/30 01:38:59 term Node 1 <Nelgin> YT: EMUFS host file and print access available
6/30 01:38:59 term Node 1 <Nelgin> YT: dosemu CDROM driver installed (V0.2)
6/30 01:38:59 term Node 1 <Nelgin> YT: Kernel: allocated 115 Diskbuffers = 62560 Bytes in HMA
6/30 01:38:59 term Node 1 <Nelgin> YT: Process 0 starting: J:\command.com /e:384 /p /k %FDPP_AUTOEXEC%
6/30 01:38:59 term Node 1 <Nelgin> YT: FreeDOS userspace tools can be installed with: insfdusr
6/30 01:38:59 term Node 1 <Nelgin> YT: Sound not enabled in config!
6/30 01:38:59 term Node 1 <Nelgin> YT: Welcome to dosemu2!
6/30 01:38:59 term Node 1 <Nelgin> YT: Build 2.0pre9

I'm really no further along than I was 2 days ago at this point.

**Comment by nelgin** [comment]
@stsp Are you waiting on me for something?

**Comment by stsp** [comment]
Well for me local.exe only shows
this:
```
                                 Yankee Trader
                               (c)Alan Davenport
          Earth & Xannor fix, Anticloak, Genesis, Missiles disabled
 
                 Strategy Guide: www.starflt.com/yt.html
 
                     Version 3.6d * YT * Modded 3/28/2021
 
                         UNREGISTERED EVALUATION COPY!
 
              PLEASE ENCOURAGE YOUR SYSOP TO REGISTER THIS GAME.
 
YT-SUB DEBUG Error Trap Entry ERL=  0 ERR= 63
```

I tried under dosbox and got the
same error.
So now I am not even sure the
copy you pointed to, is the same
as you have...
What does this error mean?

**Comment by nelgin** [comment]
I'm attaching the YT from my BBS so you have exactly the same copy as I do with the same config.
[yt.zip](https://github.com/dosemu2/dosemu2/files/9029924/yt.zip)

It'll extract into a directory "yankee"

**Comment by stsp** [comment]
Entirely different behaviour!
Although binaries seem to be
identical...
Can you figure out the difference
(maybe in configs) that lead to
such problems of the copy you
initially pointed to?

**Comment by nelgin** [comment]
If you convert 63 to hex it becomes 3F which is "Print file deleted" so maybe it's missing a file. Did you run YTCONFIG and RMT-INIT?

**Comment by stsp** [comment]
I didn't run anything.
I ask you to unpack and try
the archive you provided to
me, yourself. And made sure
it doesn't work.

Yes, it misses some .dat
files, dorinfo1.def is different
and so on. Please figure out
the problem yourself, because
you provided an URL to archive
that you didn't even try yourself.

**Comment by nelgin** [comment]
> I ask you to unpack and try
> the archive you provided to
> me, yourself. And made sure
> it doesn't work.

Sure, let me do that now and I'll post results.

**Comment by nelgin** [comment]
Please see attached.
[yt.log](https://github.com/dosemu2/dosemu2/files/9030111/yt.log)

**Comment by stsp** [comment]
Please use `dosemu -t`

**Comment by nelgin** [comment]
Getting somewhere. 

Though its displaying stuff where it shouldn't. I guess there is a status line and it's just scrolling up. 
![image](https://user-images.githubusercontent.com/6232589/176932530-6d8b9424-c847-479e-8e16-e1c862af48c0.png)

bbs        90710   90578  0 16:16 pts/3    00:00:00 /bin/bash /usr/bin/dosemu -t
bbs        90719   90710  2 16:16 pts/3    00:00:06 /usr/bin/dosemu.bin -o /home/bbs/.dosemu/boot.log -L Command line: /usr/bin/dosemu.bin -o /home/bbs/.dosemu/boot.log -t

So I guess I need to start again with an original dosemu.conf and start to pare it down until I find out what breaks it.

**Comment by stsp** [comment]
This all can only be visible
with your copy, not with the
copy from URL of this ticket.

**Comment by stsp** [comment]
So I don't really care what
will you figure out about your
custom dosemu.conf, but
please figure out what breaks
it in the original URL, which you
probably provided w/o trying
yourself.

**Comment by nelgin** [comment]
I already told you what breaks with the original file from the URL if you bothered to read the log I attached. 

![image](https://user-images.githubusercontent.com/6232589/176937823-f06b94e9-94b6-4cd3-833d-17be7f2e419a.png)

Isn't this the same location as I added to the ticket when I created it? 

![image](https://user-images.githubusercontent.com/6232589/176937909-8142e6b2-d029-4714-a087-4ce6350dccbc.png)


Why, yes it is.

> but
> please figure out what breaks
> it in the original URL, which you
> probably provided w/o trying
> yourself.

Which bit in the yt.log did I not try myself, exactly?

**Comment by stsp** [comment]
> Which bit in the yt.log did I not try myself, exactly?

I don't care about your log.
What is important is that you
never said yt-init should be
run to avoid an error msg.
You didn't tell it even now,
you somehow assume I should
find that out from your log and
not from a ticket you filled...

**Comment by nelgin** [comment]
Since you don't care. I don't care.

**Comment by andrewbird** [comment]
@nelgin from your log
> At this stage a SDL window opens. cut & paste doesn't work. so roughly..

Copy paste works in dosemu, but you need to press the shift key along with the usual mouse buttons. So 
shift +left mouse and drag to select region
shift + middle mouse to paste

**Comment by nelgin** [comment]
Good to know, but I'm done. Thanks anyway.

**Comment by stsp** [comment]
You simply don't understand
that by just saying that yt-init should
be run, you would make it possible
for me to run it week ago.
By not using some custom
dosemu.conf you would yourself
run at least local.exe week ago.
We spent a week and only got to
fixing an SDL bug when you
provided me with your local
copy that already had yt-init ran.
Was the week worth of 2 trivial
things: try w/o custom dosemu.conf
and tell about yt-init?
I don't think so.
Its really a must, to try all trivial
things yourself before asking others
to. Its really a must to provide all
test-cases and instructions to use
them. You either do that or we can't
help (and you chose later).

**Comment by stsp** [comment]
Hi @nelgin I've finally got around to
implement the needed functionality
for bbs doors. At the very minimum
you'll be able to see the boot process
now, while debugging things.
At a maximum - use proper setup
with new features.
So what is this.

First, you can now do:
`$_com1 = "exec rxvt -pty-fd 0"`
You can use any other terminal, I
only tried rxvt. In this case rxvt
communicates with com1, so you
see your bbs game in a separate
rxvt window. In dosemu window
you see all boot process etc, so
no more "oh this all hangs".

Second, if you don't want an extra
window, there is now a `comredir`
tool to redirect com port to a console.
To use that method, you first set up
the null-modem cable this way:
```
$_com1 = "nullmodem 2"
$_com2 = "nullmodem 1"
```
Now com1 is connected to com2.
You set your game to com1 and
use `comredir 2` to attach the console
to com2 and play the game.
Its best to put `comredir 2` into a
start-up bat file right before starting
the bbs software. If you want to
suppress the local output the way
`$_com1="virtual"` did, you can use
the `-s` switch of `comredir`.

Third, you can do `$_com1="/dev/tty"`.
This is the same as `$_com1="virtual"`
but without a local output suppression
so you can see the boot process.

**Comment by nelgin** [comment]
Interesting developments. Thank you for sharing this update.

I assume there'll be no issue running two games concurrently with any of these setups?

Thanks,
Nigel

**Comment by stsp** [comment]
How is so?
You mean by running 2 dosemu processes?
In that case it should work concurrently.
